### ISSUE TITLE
Use the new popup-based Canvas API authorization when launching Canvas Files assignments

### ISSUE THREAD

**Issue by seanh** [hypothesis/lms#622]
The new Canvas Files and OAuth 2.0 frontend and backend needs to be used when launching assignments, as well as when creating them.

1. User launches a Canvas Files assignment

2. Server renders JavaScript code, authentication JWT and any other necessary config into HTML page

3. Server does not render any Via iframe in the initial page load. Without JavaScript the user would see only a blank white page.

   If the assignment is an HTML or PDF public URL or a Google Drive assignment then the server *will* prerender the Via iframe into the initial response, so there's nothing for the JavaScript code to do. But for Canvas Files assignments the server will never render the iframe.

4. JavaScript code runs on page load and notices that there's no Via iframe, so it sends a JWT-authenticated XHR request to the `/via_url` proxy API.

5. If the API responds with an error:

   1. The same steps below apply whether the error was because we don't have an access token for the user or because of some other reason, the frontend should behave the same in all cases.

   2. The frontend code shows an authorization dialog:

      ```
      You need to authorize Hypothesis to access this assignment:

                                       <button>Authorize</button>
      ```

      For bonus points if the `/via_url` request failure was for some reason other than that we simply don't have an access token for this user, then we could add some information about the failure into the dialog:

       ```
       Authorization failed: <reason>. Try again:

                       <button>Authorize</button>
       ```

   3. When the user clicks the <kbd>Authorize</kbd> button the popup window opens and goes to the Canvas authorize page.

   4. When the popup window closes for any reason the frontend code tries the Via URL proxy API request again.

   5. If the request succeeds this time continue to step 6 below. If the request failed to back to ii above.

6. If the API responds successfully with the Via URL then use it to inject Via iframe into DOM

**Comment by seanh** [comment]
@robertknight FYI I've tweaked the steps above from what I first described. I think the server should make the decision about whether the Via `<iframe>` can be rendered in the initial page load, or whether we instead need to render a message asking the user to click <kbd>Authorize</kbd> and open the popup window. JavaScript code is only involved in this second case, and the JavaScript code doesn't need to decide whether it should get involved or not -- the server will either include the authorization message and JavaScript code in the page or not, as appropriate.

This actually looks pretty simple now

**Comment by seanh** [comment]
I'll prioritise the backend parts of this first next week to unblock the frontend

**Comment by seanh** [comment]
Added a checklist above

**Comment by seanh** [comment]
Thinking about the design of the download URL proxy API. Here's the Canvas API that it wraps, it's called the "inline preview" API in the docs but the URL says `public_url` and this is what the app currently uses to get a download URL: https://canvas.instructure.com/doc/api/files.html#method.files.public_url

```
GET /api/v1/files/:id/public_url
```

Response:

```json
{ "public_url": "https://example-bucket.s3.amazonaws.com/example-namespace/attachments/1/example-filename?AWSAccessKeyId=example-key&Expires=1400000000&Signature=example-signature" }
```

So the only parameter that's required is the file ID and it looks like these file IDs are global within the Canvas instance as no course ID param is given in the request.

The backend will need to make sure that this file ID, or perhaps just the whole proxy API URL to get, is in the JS config object.

Here's what I'm thinking for the proxy API:

```
GET https://lms.hypothes.is/api/canvas/files/<ID>/public_url Authorization: Bearer <JWT>
-> 200 OK {"public_url": "https://..."}
```

We'll have to decide on an error response format with error message later too.

**Comment by seanh** [comment]
Fake version of the public URL API: https://github.com/hypothesis/lms/pull/631

**Comment by seanh** [comment]
Real version of the API: https://github.com/hypothesis/lms/pull/634

**Comment by seanh** [comment]
I think we need to tweak the approach here slightly because we don't want the backend to make any synchronous requests to the Canvas API (e.g. to get a public URL for the Canvas file, or to refresh the access token) during responding to the launch request. All Canvas API requests should be done by JavaScript calling our Canvas proxy API.

The **unhappy path** will remain as previously described (server does not render Via iframe into page, JavaScript has to call `/via_url` proxy API to get the Via URL and then inject the iframe into the DOM). But the **happy path** has now changed to be the same as the unhappy path -- unlike previously described, in the happy path the server will _not_ pre-render the Via iframe into the HTML, the JavaScript still has to call the `/via_url` API and inject the iframe.

I'll update the PR description with this new plan as well:

1. User launches a Canvas Files assignment

2. Server renders JavaScript code, authentication JWT and any other necessary config into HTML page

3. Server does not render any Via iframe in the initial page load. Without JavaScript the user would see only a blank white page.

   If the assignment is an HTML or PDF public URL or a Google Drive assignment then the server *will* prerender the Via iframe into the initial response, so there's nothing for the JavaScript code to do. But for Canvas Files assignments the server will never render the iframe.

4. JavaScript code runs on page load and notices that there's no Via iframe, so it sends a JWT-authenticated XHR request to the `/via_url` proxy API.

5. If the API responds with an error:

   1. The same steps below apply whether the error was because we don't have an access token for the user or because of some other reason, the frontend should behave the same in all cases.

   2. The frontend code shows an authorization dialog:

      ```
      You need to authorize Hypothesis to access this assignment:

                                       <button>Authorize</button>
      ```

      For bonus points if the `/via_url` request failure was for some reason other than that we simply don't have an access token for this user, then we could add some information about the failure into the dialog:

       ```
       Authorization failed: <reason>. Try again:

                       <button>Authorize</button>
       ```

   3. When the user clicks the <kbd>Authorize</kbd> button the popup window opens and goes to the Canvas authorize page.

   4. When the popup window closes for any reason the frontend code tries the Via URL proxy API request again.

   5. If the request succeeds this time continue to step 6 below. If the request failed to back to ii above.

6. If the API responds successfully with the Via URL then use it to inject Via iframe into DOM

**Comment by robertknight** [comment]
Implemented in https://github.com/hypothesis/lms/pull/745.
